### PR TITLE
scsm-88 Support Dynamic Binding Channel Limits

### PR DESCRIPTION
**Pull request by garyrussell**
Required by scsm-88.

If `outputChannelNames` is not zero length, only dynamically bind if the channel is is in the list.

### CODE REVIEW

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 1]
```diff
@@ -49,7 +49,7 @@ public MessageChannel resolveDestination(String name) {
 		}
 		catch (DestinationResolutionException e) {
 		}
-		if (name.contains(":")) {
+		if (name.contains(":")) { // TODO : optional - only needed if transport provided
 			if (binderFactory != null) {
 				String[] tokens = name.split(":", 2);
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Must change the maximum to 3. Will fix on merge.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 2]
```diff
@@ -66,14 +66,11 @@ else if (tokens.length == 3) {
 							" [<transport>:]<type>:<name>");
 				}
 				Binder<MessageChannel> binder = binderFactory.getBinder(transport);
-				if ("queue".equals(type)) {
+				if ("queue".equals(type)) { // TODO: remove
 					channel = binder.bindDynamicProducer(name, this.producerProperties);
 				}
-				else if ("topic".equals(type)) {
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Ditching the 'topic' prefix makes it a little ambiguous, as '[transport]:[name]' and 'queue:[name]' will take the branch where the `type` becomes `[transport]`. IMO should add another check that the first token is not `queue` before deciding which scheme are we going to use on parsing. Of course, that will cause issues with folks naming a transport `queue`, but that's fine, I would suppose. And `queue` is rather not long lived anyway.

**Comment by mbogoevici** [comment]
LGTM - with the two notes above, which can be both fixed on merge. 
Could also use a unit test for multibinder.

**Comment by garyrussell** [comment]
Thanks, @mbogoevici , I am going to rework this as a PoC based on the slack discussion I had with @markfisher this morning (white list of allowable dynamic channels). I will add test case(s).

**Comment by artembilan** [comment]
Looks like you are talking here about https://jira.spring.io/browse/INT-1926

**Comment by garyrussell** [comment]
Yes; we should probably go ahead and implement that, in any case, but here it's a bit more complicated than that.

Consider: `expression="headers['foo']"` to select the channel name.

In this case, we can't do it in the router; the router has no information about the binding properties; we need to make available the allowable channel names to the binder aware channel resolver so it will only dynamically bind if the channel is in the white list (or there are no restrictions).

In other words, if the binding properties contains an outputChannel list, only they are allowed; if no list is provided the resolver needs to bind the channels dynamically.

i.e. the decision has to be made in the resolver, not in the router.

Let me hack together a straw-man and we can discuss further.

**Comment by garyrussell** [comment]
Pushed

**Comment by garyrussell** [comment]
Updated for the newly refactored `Binder`; ready for review, but please don't merge until I rework the router PR in s-c-s-m.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 3]
```diff
@@ -86,8 +93,22 @@ else if (tokens.length != 1) {
 								" [<transport>:]<name>");
 					}
 				}
-				Binder<MessageChannel> binder = binderFactory.getBinder(transport);
-				binder.bindProducer(name, channel, this.producerProperties);
+				String[] outboundChannelNames = null;
+				Properties producerProperties = null;
+				if (this.channelBindingServiceProperties != null) {
+					outboundChannelNames = this.channelBindingServiceProperties.getOutboundChannelNames();
+					// TODO: need the props to return some defaults if not found
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: As mentioned on Slack, I strongly believe we should restore global producer and consumer properties in `ChannelBindingServiceProperties`

Comment (garyrussell) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: +1 :+1:

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java [hunk 4]
```diff
@@ -50,6 +50,8 @@
 
 	private String defaultBinder;
 
+	private String[] outboundChannelNames = new String[0];
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java]: This property might be slightly confusing, since it does not include the outputs of `Source`, `Processor` or anything else specified via `@EnableBinding`, which definitely *are* output channels (granted, this feature is intended to be used mainly by Sinks). This is very specific to dynamic binding so I think we should find a way to express that.

Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java]: Also we're not strictly talking about "channels" here, but "destinations". Perhaps the property could be renamed something like `routableDestinations`?

Comment (garyrussell) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java]: How about `allowedDynamicDestinations` ?

Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java]: +1 on @garyrussell 's suggestion

Comment (markfisher) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/ChannelBindingServiceProperties.java]: sounds good - I like the inclusion of "dynamic"

**Comment by garyrussell** [comment]
Pushed testing support; I went with `dynamicDestinations` for now; we can talk more.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 5]
```diff
@@ -86,8 +93,23 @@ else if (tokens.length != 1) {
...
-				Binder<MessageChannel> binder = binderFactory.getBinder(transport);
-				binder.bindProducer(name, channel, this.producerProperties);
+				String[] dynamicDestinations = null;
+				Properties producerProperties = null;
+				if (this.channelBindingServiceProperties != null) {
+					dynamicDestinations = this.channelBindingServiceProperties.getDynamicDestinations();
+					// TODO: need the props to return some defaults if not found
+					producerProperties = this.channelBindingServiceProperties.getProducerProperties(name);
+				}
+				boolean dynamicAllowed = dynamicDestinations == null
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Possibly simplify as o.s.u.ObjectUtils.isEmpty()

File: spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binding/ChannelBindingServiceTests.java [hunk 6]
```diff
@@ -94,10 +110,51 @@ public void testExplicitGroup() throws Exception {
...
 		Binding<MessageChannel> mockBinding = Binding.forConsumer("foo", "fooGroup", Mockito.mock(AbstractEndpoint.class),
 				inputChannel, null);
-		Mockito.when(binder.bindConsumer("foo", "fooGroup", inputChannel, new Properties()))
+		when(binder.bindConsumer("foo", "fooGroup", inputChannel, new Properties()))
 				.thenReturn(mockBinding);
 		Binding<MessageChannel> binding = service.bindConsumer(inputChannel, name);
-		Assert.assertThat(binding, sameInstance(mockBinding));
+		assertThat(binding, sameInstance(mockBinding));
+
+		final AtomicReference<MessageChannel> dynamic = new AtomicReference<>();
```
Comment (mbogoevici) [inline spring-cloud-stream/src/test/java/org/springframework/cloud/stream/binding/ChannelBindingServiceTests.java]: I'd make the second part a separate test, just to avoid a possible confusion. Also because it reuses the mockBinding.

**Comment by mbogoevici** [comment]
These are my only observations, LGTM in general. Can fix on merge if agreed.

**Comment by garyrussell** [comment]
OK; thanks; sounds good

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 7]
```diff
@@ -40,36 +42,41 @@
...
 		MessageChannel channel = null;
+		DestinationResolutionException destinationResolutionException;
 		try {
 			return super.resolveDestination(name);
 		}
 		catch (DestinationResolutionException e) {
+			destinationResolutionException = e;
 		}
 		synchronized (this) {
 			try {
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Seems like the superclass call is duplicate, once inside the synchronized block, and once outside. Is this a double-check locking?

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 8]
```diff
@@ -86,8 +93,23 @@ else if (tokens.length != 1) {
...
+				Properties producerProperties = null;
+				if (this.channelBindingServiceProperties != null) {
+					dynamicDestinations = this.channelBindingServiceProperties.getDynamicDestinations();
+					// TODO: need the props to return some defaults if not found
+					producerProperties = this.channelBindingServiceProperties.getProducerProperties(name);
+				}
+				boolean dynamicAllowed = dynamicDestinations == null
+						|| dynamicDestinations.length == 0
+						|| Arrays.asList(dynamicDestinations).contains(name);
+				if (dynamicAllowed) {
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Also, should we do this check before registering the channel? It seems like there would be little reason to do the channel registration unless we can actually bind to it.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java [hunk 9]
```diff
@@ -86,8 +93,23 @@ else if (tokens.length != 1) {
...
+					Binder<MessageChannel> binder = binderFactory.getBinder(transport);
+					binder.bindProducer(name, channel, producerProperties);
```
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: While reviewing this, I realized that we never unbind these channels. While the endpoints are in principle stopped eventually, I still think it'd be preferable if the bindings would be part of the managed input/output binding lifecycle.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java (hunk 7 above)
Comment (mbogoevici) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BinderAwareChannelResolver.java]: Actually it is. IntelliJ threw me off I guess.

**Comment by mbogoevici** [comment]
Minor polishing here: https://github.com/mbogoevici/spring-cloud-stream/commit/fbb51771741a4af40f935e3ebdbaea795350a3c0 

(Does not address unbinding of dynamic channels, that can be done separately)

**Comment by garyrussell** [comment]
Polishing looks good; thanks; yes the double-check locking is to avoid synchronizing once the channel has been registered with the context.

I thought about unbinding - we might be able to add something in the router when it's stopped.

**Comment by mbogoevici** [comment]
Merged w/https://github.com/spring-cloud/spring-cloud-stream/commit/16e7893e497ddf2ab03c229545e4326230c73dc9, additional minor polishing https://github.com/spring-cloud/spring-cloud-stream/commit/2f330c7bc477c65bd2dc571a5507164eca3d3bc3
